core/services/ardupilot_manager/firmware/FirmwareManagement.py

```diff
@@ -0,0 +1,82 @@
+)
+from firmware.FirmwareInstall import FirmwareInstaller
+
+
+class FirmwareManager:
+    def __init__(self, firmware_folder: pathlib.Path) -> None:
+        self.firmware_folder = firmware_folder
+        self.firmware_download = FirmwareDownloader()
+        self.firmware_installer = FirmwareInstaller(self.firmware_folder)
+
+    @staticmethod
+    def firmware_name(platform: Platform) -> str:
+        """Get consistent firmware name for given platform."""
+        return f"ardupilot_{platform.value.lower()}"
+
+    def firmware_path(self, platform: Platform) -> pathlib.Path:
+        """Get firmware's path for given platform."""
+        return pathlib.Path.joinpath(self.firmware_folder, self.firmware_name(platform))
+
+    def is_firmware_installed(self, platform: Platform) -> bool:
+        """Check if firmware for given platform is installed."""
+        if platform == Platform.Undefined:
+            raise UndefinedPlatform("Platform is undefined. Cannot verify if firmware is installed.")
+
+        if platform == Platform.Pixhawk1:
+            # Assumes for now that a Pixhawk always has a firmware installed, which is true most of the time
+            # TODO: Validate if properly. The uploader tool seems capable of doing this.
+            return True
+
+        firmware_format = FirmwareDownloader._supported_firmware_formats[platform]
+        if firmware_format == FirmwareFormat.ELF:
+            return pathlib.Path.is_file(self.firmware_path(platform))
+
+        raise UnsupportedPlatform("Install check is not implemented for this platform.")
+
+    def get_available_firmwares(self, vehicle: Vehicle, platform: Platform) -> Dict[str, str]:
+        firmwares = {}
+        versions = self.firmware_download.get_available_versions(vehicle, platform)
+        if not versions:
+            raise NoVersionAvailable(f"Failed to find any version for vehicle {vehicle}.")
+        for version in versions:
+            try:
+                url = self.firmware_download.get_download_url(vehicle, platform, version)
+                firmwares[version] = url
+            except Exception as error:
+                logger.debug(f"Error fetching URL for version {version} on vehicle {vehicle}: {error}")
+        if not firmwares:
+            raise NoVersionAvailable(f"Failed do get any valid URL for vehicle {vehicle}.")
+        return firmwares
+
+    def install_firmware_from_file(self, new_firmware_path: pathlib.Path, platform: Platform) -> None:
+        try:
+            self.firmware_installer.install_firmware(new_firmware_path, self.firmware_path(platform), platform)
+            logger.info(f"Succefully installed firmware for {platform}.")
+        except Exception as error:
+            error_message = f"Could not install firmware: {error}"
+            logger.exception(error_message)
+            raise FirmwareInstallFail(error_message) from error
+
+    def install_firmware_from_url(self, url: str, platform: Platform) -> None:
```

Comment:
there is no validation for anything ?

Reply:
When you call `install_firmware` it calls the validation method.

Reply:
Regarding validation of ELF firmwares, the old approach (checking the return code on the `--help` command) is not compatible with stable SITL firmwares, since they use to return 1 on `--help` (yep).

The perfect approach would be to use the firmware validation tool you've added to Ardupilot, but it's currently broken for SITL firmwares.

For now I will leave this unvalidated and opened issue #430 to track that.

Reply:
Also added a return message stating no firmware validation is happening in the case of SITL or Navigator.